modules/api/src/main/java/org/apache/ignite/table/QualifiedName.java

```diff
@@ -0,0 +1,159 @@
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Class represents a catalog object name (table, index and etc.) and provides factory methods.
+ *
+ * <p>Qualified name is a pair of schema name and object name. Schema name is optional, when not set, then
+ * default schema name "PUBLIC" is used.
+ *
+ * <p>Both schema name and object name can be case sensitive or insensitive respecting SQL-parser style quotation.
+ * Unquoted name will be cast to upper case. E.g. "tbl0" - is equivalent to "TBL0", "\"Tbl0\"" - "Tbl0", etc.
+ */
+public class QualifiedName {
+    static final String DEFAULT_SCHEMA_NAME = "PUBLIC";
+
+    /**
+     * Parses given simple or canonical name, and returns the object qualified name.
+     */
+    public static QualifiedName parse(String simpleOrCanonicalName) {
+        verifyObjectIdentifier(simpleOrCanonicalName);
+
+        Tokenizer tokenizer = new Tokenizer(simpleOrCanonicalName);
+
+        String schemaName = DEFAULT_SCHEMA_NAME;
+        String objectName = tokenizer.nextToken();
+
+        if (tokenizer.hasNext()) {
+            // Canonical name
+            schemaName = objectName;
+            objectName = tokenizer.nextToken();
+        }
+
+        if (tokenizer.hasNext()) {
+            throw new IllegalArgumentException("Canonical name format mismatch: " + simpleOrCanonicalName);
+        }
+
+        verifySchemaIdentifier(schemaName);
+        verifyObjectIdentifier(objectName);
+
+        return new QualifiedName(schemaName, objectName);
+    }
+
+    /**
+     * Resolves given simple name against default schema.
+     */
+    public static QualifiedName fromSimple(String simpleName) {
+        return of(null, simpleName);
+    }
+
+    /**
+     * Normalize schemaName and objectName, and returns the object qualified name.
+     *
+     * @param schemaName Schema name or {@code null} for default schema.
+     * @param objectName Object name.
+     */
+    public static QualifiedName of(@Nullable String schemaName, String objectName) {
+        String schemaIdentifier = schemaName == null ? DEFAULT_SCHEMA_NAME : IgniteNameUtils.parseSimpleName(schemaName);
+        String objectIdentifier = IgniteNameUtils.parseSimpleName(objectName);
+
+        verifySchemaIdentifier(schemaIdentifier);
+        verifyObjectIdentifier(objectIdentifier);
+
+        return new QualifiedName(schemaIdentifier, objectIdentifier);
+    }
+
+    private static void verifyObjectIdentifier(@Nullable String identifier) {
+        Objects.requireNonNull(identifier);
+
+        if (identifier.isEmpty()) {
+            throw new IllegalArgumentException("Object identifier can't be empty.");
+        }
+    }
+
+    private static void verifySchemaIdentifier(@Nullable String identifier) {
+        if (identifier != null && identifier.isEmpty()) {
+            throw new IllegalArgumentException("Schema identifier can't be empty.");
+        }
+    }
+
+    /** Normalized schema name. */
+    private final String schemaIdentifier;
```

Comment:
plz move it into beginning, just become more readable

Reply:
done.